src/Lean/LocalContext.lean

```diff
@@ -48,10 +48,33 @@ See `LocalDecl.index`, `LocalDecl.fvarId`, `LocalDecl.userName`, `LocalDecl.type
 arguments common to both constructors.
 -/
 inductive LocalDecl where
-  /-- A local variable. -/
+  /-- A local variable without any value.
+  `Lean.LocalContext.mkBinding` creates lambdas or foralls from `cdecl`s. -/
   | cdecl (index : Nat) (fvarId : FVarId) (userName : Name) (type : Expr) (bi : BinderInfo) (kind : LocalDeclKind)
-  /-- A let-bound free variable, with a `value : Expr`. -/
-  | ldecl (index : Nat) (fvarId : FVarId) (userName : Name) (type : Expr) (value : Expr) (nonDep : Bool) (kind : LocalDeclKind)
+  /-- A let-bound free variable, with a value `value : Expr`.
+  If `nondep := false`, then the variable is definitionally equal to its value.
+  If `nondep := true`, then the variable has an opaque value; we call these "have-bound free variables."
+  `Lean.LocalContext.mkBinding` creates let/have expressions from `ldecl`s.
+
+  **Important:** The `nondep := true` case is subtle; it is not merely an opaque `ldecl`!
+  - In most contexts, nondependent `ldecl`s should be treated like `cdecl`s.
+    For example, suppose we have a tactic goal `x : α := v (nondep) ⊢ b`.
+    It would be incorrect for `revert x` to produce the goal `⊢ have x : α := v; b`,
+    since this would be saying "to prove `b` without knowledge of the value of `x`, it suffices to
+    prove `have x : α := v; b` for this particular value of `x`."
+    Instead, `revert x` *must* produce the goal `⊢ ∀ x : α, b`.
+    Furthermore, given a goal `⊢ have x : α := v; b`, the `intro x` tactic should yield a *dependent* `ldecl`,
+    since users expect to be able to make use of the value of `x`,
+    and also the value creates a hidden source of dependencies on other local variables.
+  - Also: `value` might not be type correct. Metaprograms may decide to pretend that all `nondep := true`
+    `ldecl`s are `cdecl`s (for example, when reverting variables). As a consequence, nondep `ldecl`s may
```

Comment:
Should we have a debugging flag to force Lean to type check the value at functions such as `withLetDecl`?